File: client/src/test/scala/stasis/test/specs/unit/client/compression/GzipSpec.scala
package stasis.test.specs.unit.client.compression

import org.apache.pekko.actor.ActorSystem
import org.apache.pekko.stream.scaladsl.Source
import org.apache.pekko.util.ByteString

import stasis.client.compression.Gzip
import stasis.test.specs.unit.AsyncUnitSpec
import stasis.test.specs.unit.client.EncodingHelpers

class GzipSpec extends AsyncUnitSpec with EncodingHelpers {
  "A Gzip encoder/decoder implementation" should "provide its name" in {
    Gzip.name should be("gzip")
  }

  it should "compress data" in {
    Source
      .single(ByteString(decompressedData))
      .via(Gzip.compress)
      .runFold(ByteString.empty)(_ concat _)
      .map { actualCompressedData =>
        actualCompressedData should be(compressedData.decodeFromBase64)
      }
  }

  it should "decompress data" in {
    Source
      .single(compressedData.decodeFromBase64)
      .via(Gzip.decompress)
      .runFold(ByteString.empty)(_ concat _)
      .map { actualDecompressedData =>
        actualDecompressedData.utf8String should be(decompressedData)
      }
  }

  private implicit val system: ActorSystem = ActorSystem(name = "GzipSpec")

  private val decompressedData = "some-decompressed-data"
  private val compressedData = "H4sIAAAAAAAAACrOz03VTUlNzs8tKEotLk5N0U1JLEkEAAAA//8DAG894xUWAAAA"
}

File: client-android/lib/src/test/kotlin/stasis/test/client_android/lib/compression/GzipSpec.kt
package stasis.test.client_android.lib.compression

import io.kotest.core.spec.style.WordSpec
import io.kotest.matchers.collections.shouldBeIn
import io.kotest.matchers.shouldBe
import okio.Buffer
import okio.ByteString.Companion.decodeBase64
import okio.buffer
import stasis.client_android.lib.compression.Gzip

class GzipSpec : WordSpec({
    "A Gzip encoder/decoder implementation" should {
        val decompressedData = "some-decompressed-data"
        val compressedData11 = "H4sIAAAAAAAAACrOz03VTUlNzs8tKEotLk5N0U1JLEkEAAAA//8DAG894xUWAAAA" // jdk11
        val compressedData17 = "H4sIAAAAAAAA/yrOz03VTUlNzs8tKEotLk5N0U1JLEkEAAAA//8DAG894xUWAAAA" // jdk17

        "provide its name" {
            Gzip.name() shouldBe ("gzip")
        }

        "compress data" {
            val source = Buffer().write(decompressedData.toByteArray())

            val actualCompressedData = Gzip.compress(source).buffer().use {
                it.readByteString()
            }

            actualCompressedData.base64() shouldBeIn (listOf(compressedData11, compressedData17))
        }

        "decompress data" {
            val source11 = Buffer().write(compressedData11.decodeBase64()!!)
            val source17 = Buffer().write(compressedData17.decodeBase64()!!)

            val actualDecompressedData11 = Gzip.decompress(source11).buffer().readUtf8()
            actualDecompressedData11 shouldBe (decompressedData)

            val actualDecompressedData17 = Gzip.decompress(source17).buffer().readUtf8()
            actualDecompressedData17 shouldBe (decompressedData)
        }
    }
})
